File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/handlers/WritableHealthRecordHandler.kt
package com.phamtunglam.health_connector_hc_android.handlers

import android.os.RemoteException
import com.phamtunglam.health_connector_hc_android.HealthConnectorClient
import com.phamtunglam.health_connector_hc_android.logger.HealthConnectorLogger
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.HealthRecordDto
import java.io.IOException

/**
 * Capability protocol for handlers that support writing health records to Health Connect.
 *
 * This interface provides a default implementation for writing individual health records.
 * Batch write operations are handled atomically at the [HealthConnectorClient]
 * level to ensure all-or-nothing semantics across different record types.
 */
internal interface WritableHealthRecordHandler : HealthRecordHandler {
    /**
     * Writes a single health record to Health Connect.
     *
     * This method converts the DTO to a Health Connect record and inserts it atomically.
     *
     * @param dto The health record DTO to write
     * @return The platform-assigned record ID string
     *
     * @throws RemoteException if the Health Connect service is unreachable
     * @throws SecurityException if write permission is not granted
     * @throws IOException if a network or storage error occurs
     */
    @Throws(
        RemoteException::class,
        SecurityException::class,
        IOException::class,
    )
    suspend fun writeRecord(dto: HealthRecordDto): String {
        val operation = "write_record"
        val context = mapOf("data_type" to dataType.name)

        return process(
            operation = operation,
            context = context,
        ) {
            HealthConnectorLogger.debug(
                tag = tag,
                operation = operation,
                message = "Preparing to write record",
                context = context,
            )

            val record = dto.toHealthConnect()

            val response = client.insertRecords(listOf(record))

            val recordId = response.recordIdsList.first()

            HealthConnectorLogger.info(
                tag = tag,
                operation = operation,
                message = "Record written successfully",
                context = context,
            )

            recordId
        }
    }
}

File: packages/health_connector_hk_ios/ios/health_connector_hk_ios/Sources/health_connector_hk_ios/handlers/WritableHealthRecordHandler.swift
import Foundation
import HealthKit

/// Capability protocol for handlers that support writing health records to HealthKit.
///
/// This protocol declares `writeRecord(_:)` as a requirement so that calls made
/// through a `WritableHealthRecordHandler`-typed reference use dynamic dispatch.
/// A default implementation is provided in the protocol extension below for
/// standard record types that can be written via `HealthRecordDto.toHKSample()`.
///
/// Batch write operations are handled atomically at the `HealthConnectorClient`
/// level to ensure all-or-nothing semantics across different record types.
protocol WritableHealthRecordHandler: HealthRecordHandler {
    /// Writes a single health record to HealthKit.
    ///
    /// Conforming types can override this requirement to customize write behavior
    /// for special HealthKit types (e.g., heartbeat series) while still benefiting
    /// from the default implementation provided in the protocol extension.
    ///
    /// - Parameter dto: The health record DTO to write. Must match this handler's
    ///                  `RecordDto` associated type.
    /// - Returns: The UUID string of the created HealthKit sample.
    ///
    /// - Throws: `HealthConnectorError.invalidArgument` if the DTO type doesn't match
    ///           the handler's expected type.
    /// - Throws: `HealthConnectorError` if the HealthKit save operation fails.
    func writeRecord(_ dto: HealthRecordDto) async throws -> String
}

extension WritableHealthRecordHandler {
    /// Writes a single health record to HealthKit.
    ///
    /// - Parameter dto: The health record DTO to write. Must match this handler's
    ///                  `RecordDto` associated type.
    /// - Returns: The UUID string of the created HealthKit sample.
    ///
    /// - Throws: `HealthConnectorError.invalidArgument` if the DTO type doesn't match
    ///          the handler's expected type.
    /// - Throws: `HealthConnectorError` if the HealthKit save operation fails.
    func writeRecord(_ dto: HealthRecordDto) async throws -> String {
        // Validate that the DTO is of the expected type for this handler
        guard dto is Self.RecordDto else {
            throw HealthConnectorError.invalidArgument(
                message: "Type mismatch: expected \(Self.RecordDto.self), but got \(type(of: dto))",
                context: [
                    "expected_type": String(describing: Self.RecordDto.self),
                    "actual_type": String(describing: type(of: dto)),
                    "handler_type": String(describing: type(of: self)),
                ]
            )
        }

        let tag = String(describing: type(of: self))
        let operation = "write_record"
        let context: [String: Any] = ["data_type": Self.dataType.rawValue]

        return try await process(operation: operation, context: context) {
            HealthConnectorLogger.debug(
                tag: tag,
                operation: operation,
                message: "Preparing to write record",
                context: context
            )

            let sample = try dto.toHKSample()
            try await self.healthStore.save(sample)
            let recordId = sample.uuid.uuidString

            HealthConnectorLogger.info(
                tag: tag,
                operation: operation,
                message: "Record written successfully",
                context: context
            )

            return recordId
        }
    }
}
